sage/group_prover.sage

```diff
@@ -0,0 +1,303 @@
+      self.nonzero = dict()
+
+  def negate(self):
+    return constraints(zero=self.nonzero, nonzero=self.zero)
+
+  def __add__(self, other):
+    zero = self.zero.copy()
+    zero.update(other.zero)
+    nonzero = self.nonzero.copy()
+    nonzero.update(other.nonzero)
+    return constraints(zero=zero, nonzero=nonzero)
+
+  def __str__(self):
+    return "constraints(zero=%s,nonzero=%s)" % (self.zero, self.nonzero)
+
+  def __repr__(self):
+    return "%s" % self
+
+
+def conflicts(R, con):
+  """Check whether any of the passed non-zero assumptions is implied by the zero assumptions"""
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), con.zero))
+  if len(con.nonzero) > 0 and 1 in zero:
+    return True
+  for nonzero in con.nonzero:
+    if nonzero.iszero(zero):
+      return True
+  if reduce(lambda a,b: a * b, con.nonzero, fastfrac(R, 1)).iszero(zero):
+    return True
+  return False
+
+
+def get_nonzero_set(R, assume):
+  """Calculate a simple set of nonzero expressions"""
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), assume.zero))
+  nonzero = set()
+  for nz in assume.nonzero:
+    if mynumerator(R, nz) != 0:
+      for (f,n) in mynumerator(R, nz).factor():
+        nonzero.add(f)
+    rnz = zero.reduce(mynumerator(R, nz))
+    if rnz != 0:
+      for (f,n) in rnz.factor():
+        nonzero.add(f)
+  return nonzero
+
+
+def prove_nonzero(R, exprs, assume):
+  """Check whether an expression is provably nonzero, given assumptions"""
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), assume.zero))
+  nonzero = get_nonzero_set(R, assume)
+  expl = set()
+  ok = True
+  for expr in exprs:
+    if mynumerator(R, expr) in zero:
+      return (False, [exprs[expr]])
+  allexprs = reduce(lambda a,b: a*b, exprs, fastfrac(R, 1))
+  for (f, n) in mynumerator(R, allexprs).factor():
+    if f not in nonzero:
+      ok = False
+  if ok:
+    return (True, None)
+  ok = True
+  for (f, n) in zero.reduce(mynumerator(R, allexprs)).factor():
+    if f not in nonzero:
+      ok = False
+  if ok:
+    return (True, None)
+  ok = True
+  for expr in exprs:
+    for (f,n) in mynumerator(R, expr).factor():
+      if f not in nonzero:
+        ok = False
+  if ok:
+    return (True, None)
+  ok = True
+  for expr in exprs:
+    for (f,n) in zero.reduce(mynumerator(R, expr)).factor():
+      if f not in nonzero:
+        expl.add(exprs[expr])
+  if expl:
+    return (False, list(expl))
+  else:
+    return (True, None)
+
+
+def prove_zero(R, exprs, assume):
+  """Check whether all of the passed expressions are provably zero, given assumptions"""
+  r, e = prove_nonzero(R, dict(map(lambda x: (fastfrac(R, x.bot, 1), exprs[x]), exprs)), assume)
+  if not r:
+    return (False, map(lambda x: "Possibly zero denominator: %s" % x, e))
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), assume.zero))
+  nonzero = reduce((lambda x, y: x * y), assume.nonzero, fastfrac(R, 1))
+  expl = []
+  for expr in exprs:
+    if not (expr * nonzero).iszero(zero):
```

Comment:
There was a reason for this, but it seems it's no longer needed. Gone in next push.